### ISSUE TITLE
Using enum as paramter in fuzz test fails

### ISSUE THREAD

**Issue by david-k** [gakonst/foundry#871]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [X] Foundryup

### What version of Foundry are you on?

forge 0.1.0 (39b6e39 2022-03-08T00:05:45.429338818+00:00)

### Operating System

Linux

### Describe the bug

When using an `enum` as a parameter in a fuzz test the test will very likely fail because the provided value is invalid for the `enum`:
```solidity
contract MyTest is DSTest {
	enum Enum { A, B }

	// Always fails
	function test_enum(Enum e) public {

	}
}
```
My understanding is that this happens because enums are represented as `unit8` in the EVM, and when fuzzing we naturally want to sample values from the full range of `uint8`. However, in the case of enums this causes the test to immediately fail.

I don't know if it is possible to somehow provide additional information to the test runner so that it knows that only a subset of values should be considered.

You can of course easily work around this issue, but being able to use enums would still be nice.

**Comment by gakonst** [comment]
Ah, this seems hard to get around. The best way to do this would be:

```solidity
function test_enum(uint8 _e) public {
    vm.assume(_e < 2);
    Enum e = Enum(_e);
}
```

I don't think we can solve this in some other way, unless the compiler gives us information about how many enum variants there are? cc @mattsse

**Comment by onbjerg** [comment]
Read over the Solidity docs and it does not seem like the JSON output has any option to return info on enums

**Comment by david-k** [comment]
Yeah, I figured that this isn't easy to do. Fortunately, one can easily work around this with some modulo operations or `vm.assume()`. (Thanks btw for mentioning `vm.assume()`, didn't know about that one!)

**Comment by 0xalpharush** [comment]
I'd appreciate reconsidering how the fuzzer treats enums as it requires a lot of boilerplate to craft more complex types e.g. structs and cast unsigned integers to an enum type. 

If that's not realistic, it may be worth displaying suggestions to users when an enum is in a function signature and it's reverting. It wasn't readily apparent that the failure is due to invalid calldata.

**Comment by onbjerg** [comment]
We need solc to give us more information on enum types so we can generate inputs that are not outside of the bounds of the number of variants of the enum

**Comment by gakonst** [comment]
@0xalpharush we agree that this would be nice, as Bjerg writes above, per our last investigation, we don't have that information to provide more context to the user.. If you have any suggestions on how to do that, we'd obviously be very happy to include, as this is indeed an important shortcoming

**Comment by gakonst** [comment]
This makes sense - I think we have access to the panic code given we handle it in Forge-std. Will investigate

**Comment by hrkrshnn** [comment]
@onbjerg where do you want the extra information? In the `--abi` or `--ast-compact-json`? Can you open an issue in the solc repo?

---

As a side note, `type(Enum).max` works. So something like `vm.assume(rand <= uint8(type(Enum).max))` is probably the best right now.

**Comment by onbjerg** [comment]
We don't process the AST in `forge test` at all and I'd be hesitant to do so currently since it's not documented anywhere as far as I can tell, so implementing e.g. https://github.com/gakonst/ethers-rs/pull/1567 was a major pain (and is likely to break at some point). `vm.assume` is OK for now but having the number of variants for enums in the ABI would be nice (like we have struct descriptions)

**Comment by frangio** [comment]
I maintain documentation for the AST at https://solidity-ast.netlify.app, but dealing with different Solidity versions is difficult so to keep `forge test` AST agnostic is probably a good idea.

**Comment by Hotmanics** [comment]
Revisiting this issue as I encounter issues with my code when trying to write fuzz tests wherein a struct containing an enum throws an EvmError when the function runs.

<img width="264" alt="Screenshot 2024-03-27 at 5 12 38 PM" src="https://github.com/foundry-rs/foundry/assets/32080359/fa835f87-8aa9-4053-8ff3-f54259150ec0">

<img width="1270" alt="Screenshot 2024-03-27 at 5 13 23 PM" src="https://github.com/foundry-rs/foundry/assets/32080359/993f79a1-d594-40fc-9c58-f0d04c79a0f3">

**Comment by 0xmoebius** [comment]
> Revisiting this issue as I encounter issues with my code when trying to write fuzz tests wherein a struct containing an enum throws an EvmError when the function runs.
> 
> <img alt="Screenshot 2024-03-27 at 5 12 38 PM" width="264" src="https://private-user-images.githubusercontent.com/32080359/317532813-fa835f87-8aa9-4053-8ff3-f54259150ec0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzOTQyOTcsIm5iZiI6MTcyMTM5Mzk5NywicGF0aCI6Ii8zMjA4MDM1OS8zMTc1MzI4MTMtZmE4MzVmODctOGFhOS00MDUzLThmZjMtZjU0MjU5MTUwZWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDEyNTk1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRiYjk4ZTYwYmVlMmU5MGJhMDQyNDQxOTkyMDM2Mjg2MDRkMDM1YzEwNDBmZGY3MjBjY2NkZjE2NGE3MjVlOTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.PkcPyS4JTRcx3mXxNEeONY6ZX0oDMONTA1OKGGJGxpU"> <img alt="Screenshot 2024-03-27 at 5 13 23 PM" width="1270" src="https://private-user-images.githubusercontent.com/32080359/317532837-993f79a1-d594-40fc-9c58-f0d04c79a0f3.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEzOTQyOTcsIm5iZiI6MTcyMTM5Mzk5NywicGF0aCI6Ii8zMjA4MDM1OS8zMTc1MzI4MzctOTkzZjc5YTEtZDU5NC00MGZjLTljNTgtZjBkMDRjNzlhMGYzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE5VDEyNTk1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk1NGRhMWY2YjNiMWEyYmRmOTA2MjNiODlkMGY5ZTU5YWY4ZTljMTZmNjNjODAwNjkxNThiNmY3NjY0MTg2NmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TaUByvFi1W95ZdAb-DmGff9TVn3XEov3V5W6VQ3ezpk">

having the same issue. can't fuzz a struct w/ an enum inside because the conversion fails with any fuzzed value bigger than enum.max
